LazyProxy.Unity/UnityExtensions.cs

```diff
@@ -152,11 +160,34 @@ public static class UnityExtensions
 
             return container
                 .RegisterType(typeFrom, typeTo, registrationName, getLifetimeManager(), injectionMembers)
-                .RegisterType(typeFrom, lazyProxyType, name,
-                    getLifetimeManager(),
-                    new InjectionConstructor(
-                        new ResolvedParameter(typeof(Lazy<>).MakeGenericType(typeFrom), registrationName))
-                );
+                .RegisterType(typeFrom, name, getLifetimeManager(), new InjectionFactory(
+                    (c, t, n) =>
+                    {
+                        var valueFactory = GetValueFactory(c, t, registrationName);
+                        var lazy = Activator.CreateInstance(typeof(Lazy<>).MakeGenericType(t), valueFactory);
+
+                        var closedLazyProxyType = lazyProxyType.IsGenericTypeDefinition
+                            ? lazyProxyType.MakeGenericType(t.GenericTypeArguments)
+                            : lazyProxyType;
+
+                        return Activator.CreateInstance(closedLazyProxyType, lazy);
+                    }));
+        }
+
+        private static Delegate GetValueFactory(IUnityContainer container, Type type, string name)
+        {
+            var lazy = ValueFactories.GetOrAdd((container, type, name), tuple =>
```

Comment:
I'm not sure if we want to have `name` in key, it's randomly generated so every attempt to register something would lead to additional cache entry.

Reply:
Also, having `IUnityContainer` in key leaks it forever

Reply:
> For example, in case transient registration, it can be frequently enough. That is why caching by name is used.
`name` is randomly chosen, actual key is `type`, if we register the same type twice, this would mean two cache entries, even if they are essentially the same.

Reply:
Resolving of the real service is used concrete container and concrete name, that is why it is in key.